File: javascript/packages/linter/src/rules/actionview-no-silent-helper.ts
import { ParserRule } from "../types.js"
import { BaseRuleVisitor } from "./rule-utils.js"

import { isERBOutputNode, isPrismNodeType } from "@herb-tools/core"
import { isActionViewHelperCall } from "./action-view-utils.js"

import type { UnboundLintOffense, LintOffense, LintContext, FullRuleConfig, BaseAutofixContext, Mutable } from "../types.js"
import type { ParseResult, ERBContentNode, ParserOptions, PrismNode } from "@herb-tools/core"

interface ActionViewNoSilentHelperAutofixContext extends BaseAutofixContext {
  node: Mutable<ERBContentNode>
}

function collectDiscardedHelperCalls(prismNode: PrismNode): { helperName: string }[] {
  const matches: { helperName: string }[] = []

  const collect = (node: PrismNode | null | undefined): void => {
    if (!node) return

    const match = isActionViewHelperCall(node)

    if (match) {
      matches.push(match)
      return
    }

    if (isPrismNodeType(node, "StatementsNode")) {
      node.body.forEach(collect)
      return
    }

    if (isPrismNodeType(node, "IfNode")) {
      collect(node.statements)
      collect(node.subsequent)
      return
    }

    if (isPrismNodeType(node, "UnlessNode")) {
      collect(node.statements)
      collect(node.elseClause)
      return
    }

    if (isPrismNodeType(node, "ElseNode")) {
      collect(node.statements)
      return
    }

    if (isPrismNodeType(node, "AndNode") || isPrismNodeType(node, "OrNode")) {
      collect(node.right)
      return
    }

    if (isPrismNodeType(node, "BeginNode")) {
      collect(node.statements)
      return
    }

    if (isPrismNodeType(node, "ParenthesesNode")) {
      collect(node.body)
      return
    }
  }

  collect(prismNode)

  return matches
}

class ActionViewNoSilentHelperVisitor extends BaseRuleVisitor<ActionViewNoSilentHelperAutofixContext> {
  visitERBContentNode(node: ERBContentNode): void {
    this.checkSilentHelper(node)
    super.visitERBContentNode(node)
  }

  private checkSilentHelper(node: ERBContentNode): void {
    if (isERBOutputNode(node)) return

    const tagOpening = node.tag_opening?.value
    if (!tagOpening) return
    if (tagOpening.startsWith("<%%")) return

    const prismNode = node.prismNode
    if (!prismNode) return

    for (const match of collectDiscardedHelperCalls(prismNode)) {
      this.addOffense(
        `Avoid using \`${tagOpening} %>\` with \`${match.helperName}\`. Use \`<%= %>\` to ensure the helper's output is rendered.`,
        node.location,
        { node: node as Mutable<ERBContentNode> },
      )
    }
  }
}

export class ActionViewNoSilentHelperRule extends ParserRule<ActionViewNoSilentHelperAutofixContext> {
  static unsafeAutocorrectable = true
  static autofixRequiresContext = true
  static ruleName = "actionview-no-silent-helper"
  static introducedIn = this.version("0.9.0")

  get defaultConfig(): FullRuleConfig {
    return {
      enabled: true,
      severity: "error"
    }
  }

  get parserOptions(): Partial<ParserOptions> {
    return {
      action_view_helpers: true,
      prism_nodes: true,
    }
  }

  check(result: ParseResult, context?: Partial<LintContext>): UnboundLintOffense<ActionViewNoSilentHelperAutofixContext>[] {
    const visitor = new ActionViewNoSilentHelperVisitor(this.ruleName, context)

    visitor.visit(result.value)

    return visitor.offenses
  }

  autofix(offense: LintOffense<ActionViewNoSilentHelperAutofixContext>, result: ParseResult): ParseResult | null {
    if (!offense.autofixContext) return null

    const { node } = offense.autofixContext

    if (!node.tag_opening) return null
    if (node.tag_opening.value === "<%=") return null

    node.tag_opening.value = "<%="

    return result
  }
}

File: rust/herb-linter/src/rules/actionview_no_silent_helper.rs
use crate::utils::action_view_utils::is_action_view_helper_call;
use crate::utils::erb_utils::is_output_tag_opening;

use herb::nodes::ERBContentNode;
use herb::Visitor;

rule_visitor!(ActionViewNoSilentHelperVisitor);
define_parser_rule!(
  ActionViewNoSilentHelperRule,
  "actionview-no-silent-helper",
  Error,
  ActionViewNoSilentHelperVisitor,
  parser_options: { action_view_helpers: true, prism_nodes: true },
  introduced_in: "0.9.0"
);

impl Visitor for ActionViewNoSilentHelperVisitor {
  fn visit_erb_content_node(&mut self, node: &ERBContentNode) {
    let tag_opening = match node.tag_opening.as_ref().map(|token| token.value.as_str()) {
      Some(tag_opening) if !tag_opening.is_empty() => tag_opening,
      _ => {
        self.walk_erb_content_node(node);
        return;
      }
    };

    if is_output_tag_opening(tag_opening) || tag_opening.starts_with("<%%") {
      self.walk_erb_content_node(node);
      return;
    }

    if let Some(prism_node) = node.prism() {
      let mut helper_names = Vec::new();

      collect_discarded_helper_calls(prism_node, &mut helper_names);

      for helper_name in helper_names {
        self.add_offense(
          format!(
            "Avoid using `{} %>` with `{}`. Use `<%= %>` to ensure the helper's output is rendered.",
            tag_opening, helper_name
          ),
          node.location.clone(),
        );
      }
    }

    self.walk_erb_content_node(node);
  }
}

/// Only descends where the result of a helper is genuinely thrown away. A
/// helper whose value is assigned or passed on is consumed, so it is left alone.
fn collect_discarded_helper_calls(node: &herb::prism::PrismNode, helper_names: &mut Vec<String>) {
  if node.is("CallNode") {
    if let Some(helper_name) = is_action_view_helper_call(node) {
      helper_names.push(helper_name);
    }

    return;
  }

  let fields: &[&str] = match node.node_type.as_str() {
    "StatementsNode" => &["body"],
    "IfNode" => &["statements", "subsequent"],
    "UnlessNode" => &["statements", "else_clause"],
    "ElseNode" => &["statements"],
    "AndNode" | "OrNode" => &["right"],
    "BeginNode" => &["statements"],
    "ParenthesesNode" => &["body"],
    "ProgramNode" => &["statements"],
    _ => return,
  };

  for field in fields {
    for child in node.field(field) {
      collect_discarded_helper_calls(child, helper_names);
    }
  }
}
